naga/src/common/wgsl.rs

```diff
@@ -67,3 +67,307 @@ impl StandardFilterableTriggeringRule {
         }
     }
 }
+
+/// Types that can return the WGSL source representation of their
+/// values as a `'static` string.
+///
+/// This trait is specifically for types whose WGSL forms are simple
+/// enough that they can always be returned as a static string. IR
+/// types that require more sophisticated formatting should implement
+/// [`std::fmt::Display`] on some wrapper type.
+pub trait ToWgsl: Sized {
+    /// Return WGSL source code representation of `self`.
+    fn to_wgsl(self) -> &'static str;
+}
+
+/// Types that may be able to return the WGSL source representation
+/// for their values as a `'static' string.
+///
+/// This trait is specifically for types whose WGSL forms are simple
+/// enough that they can always be returned as a static string. IR
+/// types that require more sophisticated formatting should implement
+/// [`std::fmt::Display`] on some wrapper type.
+pub trait TryToWgsl: Sized {
```

Comment:
nitpick: It'd be nice to have these to refer to each other in docs, since they're complementary APIs.